### PR TITLE
remove CacheMeta-based caching

### PR DESCRIPTION
**Pull request by gertjanvanzwieten**
This PR removes all uses of the `__cache__` mechanism, facilitated by the `CacheMeta` meta class, and replaces it by the cached_property and `lru_cache` decorators which are considered to give better visibility to caching structures. The `CacheMeta` class is removed as no longer relevant.

### CODE REVIEW

**Comment by gertjanvanzwieten** [comment]
Oh there is actually an unresolved problem: I had to remove the base properties `Reference.vertices` and `Reference.edge_vertices` because they cannot otherwise be overridden in `MosaicReference`. @joostvanzwieten Would it be acceptable to use an annotation like `vertices: numpy.ndarray` instead?

File: nutils/element.py [hunk 1]
```diff
@@ -24,10 +25,19 @@
...
 
 class Reference(types.Singleton):
-    'reference element'
+    '''Reference element.
 
-    __slots__ = 'ndims',
-    __cache__ = 'connectivity', 'edgechildren', 'volume', 'centroid', '_linear_bernstein', 'getpoints'
+    .. py:attribute:: vertices
+
+       Array of shape `nverts x ndims` that defines the coordinates of the
```
Comment (joostvanzwieten) [inline nutils/element.py:32]: Code should be enclosed in double backticks.

File: nutils/element.py [hunk 2]
```diff
@@ -24,10 +25,19 @@
...
+       element's vertices.
+
+    .. py:attribute:: edge_vertices
+
+       Sequence of integer arrays that specifies per edge (outer sequence,
+       corresponding to `edges`) for each vertex (inner sequence, corresponding
```
Comment (joostvanzwieten) [inline nutils/element.py:38]: double backticks

File: nutils/element.py [hunk 3]
```diff
@@ -24,10 +25,19 @@
...
+       to `edges[iedge].vertices`) its index in `vertices`.
```
Comment (joostvanzwieten) [inline nutils/element.py:39]: double backticks

**Review by joostvanzwieten — APPROVED** [review]

**Review by joostvanzwieten — APPROVED** [review]

**Pull request merged**